plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Cse.hs

```diff
@@ -0,0 +1,398 @@
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TupleSections     #-}
+{-# LANGUAGE TypeApplications  #-}
+{-# LANGUAGE TypeFamilies      #-}
+{-# LANGUAGE TypeOperators     #-}
+
+module UntypedPlutusCore.Transform.Cse (cse) where
+
+import PlutusCore (MonadQuote, Name, freshName)
+import UntypedPlutusCore.Core
+import UntypedPlutusCore.Purity (isWorkFree)
+import UntypedPlutusCore.Size
+
+import Control.Lens
+import Control.Monad
+import Control.Monad.Trans.Class
+import Control.Monad.Trans.Reader
+import Control.Monad.Trans.State.Strict
+import Data.Foldable
+import Data.Hashable
+import Data.HashMap.Strict (HashMap)
+import Data.HashMap.Strict qualified as Map
+import Data.List.Extra
+import Data.Traversable
+
+{- Note [CSE]
+
+-------------------------------------------------------------------------------
+1. Simplifications
+-------------------------------------------------------------------------------
+
+This is a simplified (i.e., not fully optimal) implementation of CSE. The two simplifications
+we made are:
+
+- No alpha equivalence check, i.e., `\x -> x` and `\y -> y` are considered different expressions.
+- The builtin function arity information is approximate: rather than using the accurate arities,
+  we simply use the maximum number of arguments applied to a builtin function in the program
+  as the builtin function's arity. The arity information is used to determine whether a builtin
+  application is possibly saturated.
+
+-------------------------------------------------------------------------------
+2. How does it work?
+-------------------------------------------------------------------------------
+
+We use the following example to explain how the implementation works:
+
+\x y -> (1+(2+x))
+        +
+        (case y [ (1+(2+x)) + (3+x)
+                , (2+x) + (3+x)
+                , 4+x
+                ]
+        )
+
+The implementation makes several passes on the given term. The first pass collects builtin
+arity information as described above.
+
+In the second pass, we assign a unique ID to each `LamAbs`, `Delay`, and each `Case` branch.
```

Comment:
Why do we need one for a case branch? A case branch can be an arbitrary term, and if it's a function then there will be a lambda inside it, which you will give an id to?

Reply:
I guess I'm missing the "abstract picture". What are we trying to do here? Identify unique scopes? or what?

Reply:
Yeah, I don't know how to describe this precisely, but the purpose is to avoid evaluating something that otherwise may not be evaluated. E.g.,

- If there's a common subexpression in a parent scope and a child scope (e.g., "0.1", and "0.1.2"), then it is safe to perform CSE
- On the other hand, if it is in two scopes that don't have parent-child relationship (e.g., "0.1.2" and "0.1.3"), then it is not safe to perform CSE. Because if you do perform CSE, the new "binding" has to be in scope "0.1", but it is possible that "0.1" is evaluated without evaluating either "0.1.2" or "0.1.3".

Reply:
I think there are maybe three answers?
1. They represent places where computation stops, i.e. subexpressions are not immediately evaluated, and as such they shouldn't be hoisted above that point lest they be evaluated too soon.
2. They represent places where evaluation may happen multiple times with a different environment (lambdas) or not (delays) i.e. subexpressions may not always be evaluated to the same thing so shouldn't be shared between multiple evaluations.
3. They may bind new variables, in which case it's actually incorrect to hoist a subexpression out of one.

Reply:
I think point 1 is correct. I'll add it to the doc. I don't agree with 2 or 3: if it is incorrect to hoist something (because, e.g., it depends on the lambda-bound variable), then it can't be a common subexpression anyway.

Reply:
> if it is incorrect to hoist something (because, e.g., it depends on the lambda-bound variable), then it can't be a common subexpression anyway.

I guess the problem is the related notions of "common sub-expression" and "structurally identical term". In the presence of non-globally-unique names you might have structurally identical terms that are not in fact identical, which might require this kind of care. Are we assuming global uniqueness?

Reply:
> you might have structurally identical terms that are not in fact identical, which might require this kind of care. Are we assuming global uniqueness?

Hmm, I guess that is indeed possible. I'll make it do renaming before CSE.

Reply:
And write something!!